### PR TITLE
Redis sets

### PR DESCRIPTION
**Pull request by hackley**
This is my first stab at implementing Redis Sets. With the exception of SSCAN, all native features are available.

### CODE REVIEW

**Comment by Slava** [comment]
Thanks for this @nathanhackley!

I will look at it a bit later. Can you please talk more on the set presentation on DDP messages? Thanks for writing tests too!

**Comment by hackley** [comment]
Hey @Slava,

I'm relatively new to Meteor / the package system, could you explain what you mean by "Can you please talk more on the set presentation on DDP messages"?

Thanks!

**Comment by hackley** [comment]
I think you are referring to the reactivity of Sets, correct? I just pushed a fix that adds support for Deps – will write tests ASAP.

**Comment by Slava** [comment]
Sorry I misread the code. It doesn't touch the live-data part.

File: miniredis.js [hunk 1]
```diff
@@ -236,14 +236,15 @@ _.extend(Miniredis.RedisStore.prototype, {
...
-    if (! self._keyDependencies[key])
+    if (! self._keyDependencies[key]) {
       self._keyDependencies[key] = new Deps.Dependency();
-
-    if (Deps.active) {
-      // for future clean-up
-      Deps.onInvalidate(function () {
-        self._tryCleanUpKeyDep(key);
-      });
+    } else {
```
Comment (Slava) [inline miniredis.js:241]: can you explain this change? why did you wrap this part into an else block?

File: miniredis.js [hunk 2]
```diff
@@ -1014,3 +1015,197 @@ function patternToRegexp (pattern) {
 
 Miniredis.patternToRegexp = patternToRegexp;
 
+Miniredis.Set = function (set) {
+  this._set = set || [];
+  this._didChange = false;
+};
+
+Miniredis.Set.sunion = function (setsMembers) {
+  var flat = _.flatten(setsMembers);
```
Comment (Slava) [inline miniredis.js:1024]: I don't think this method is implemented correctly? `sunion` command gets a list of keys, and only the union of sets corresponding to those keys is returned.

File: miniredis.js [hunk 3]
```diff
@@ -1014,3 +1015,197 @@ function patternToRegexp (pattern) {
...
+  var res = _.uniq(flat);
+  return res;
+};
+
+Miniredis.Set.sdiff = function (setsMembers) {
+  var base = setsMembers.shift();
```
Comment (Slava) [inline miniredis.js:1030]: same as `sunion`, this method is implemented incorrectly.

File: miniredis.js [hunk 4]
```diff
@@ -1014,3 +1015,197 @@ function patternToRegexp (pattern) {
...
+  var sub = _.flatten(setsMembers);
+  return _.difference(base, sub);
+};
+
+Miniredis.Set.sinter = function (setsMembers) {
+  return _.intersection.apply(_, setsMembers);
```
Comment (Slava) [inline miniredis.js:1036]: same as above?

File: miniredis.js (hunk 2 above)
Comment (Slava) [inline miniredis.js:1024]: ugh, I might misread this. So this method is not on the prototype but rather on the constructor function. Can you just make it a file-scoped function?

```
var sunion = function () {...};
```

File: miniredis.js (hunk 3 above)
Comment (Slava) [inline miniredis.js:1030]: edit: same as `sunion`, make it file-scoped local.

File: miniredis.js [hunk 5]
```diff
@@ -1014,3 +1015,197 @@ function patternToRegexp (pattern) {
...
+
+_.extend(Miniredis.Set.prototype, {
+  smembers: function () {
+    return (this._set.length > 0) ? this._set : undefined;
+  },
+  sadd: function (/* values */) {
+    var self = this;
+    var values = _.invoke(arguments, "toString");
+    var insertCount = 0;
+    values.forEach(function(val) {
```
Comment (Slava) [inline miniredis.js:1047]: don't use `forEach`, it is EcmaScript5 which is not supported by some versions of IE. Instead use `_.each`.

File: miniredis.js [hunk 6]
```diff
@@ -1014,3 +1015,197 @@ function patternToRegexp (pattern) {
...
+    } else {
+      sample = _.sample(set, num);
+    }
+    return sample;
+  },
+  srem: function (/* values */) {
+    var self = this;
+    var values = _.invoke(arguments, "toString");
+    var count = 0;
+    values.forEach(function(val) {
```
Comment (Slava) [inline miniredis.js:1082]: don't use `forEach`

File: miniredis.js [hunk 7]
```diff
@@ -1014,3 +1015,197 @@ function patternToRegexp (pattern) {
...
+        sample.push(_.sample(set));
+      }
+    } else {
+      sample = _.sample(set, num);
+    }
+    return sample;
+  },
+  srem: function (/* values */) {
+    var self = this;
+    var values = _.invoke(arguments, "toString");
```
Comment (Slava) [inline miniredis.js:1080]: It is safer and more explicit to convert arguments to an array with `_.toArray` first

**Pull request closed without merging**